File: api/src/storage.ts
import { Database } from "bun:sqlite";
import { mkdir, readFile, rename } from "node:fs/promises";
import { existsSync } from "node:fs";
import { join } from "node:path";
import type { Instance, ProxyAuth, State, Subscription } from "./types";

/** 生成随机用户名（8-12 位小写字母+数字） */
function generateRandomUsername(): string {
  const chars = "abcdefghijklmnopqrstuvwxyz0123456789";
  const lenByte = new Uint8Array(1);
  crypto.getRandomValues(lenByte);
  const len = 8 + (lenByte[0] % 5); // 8-12
  let result = "";
  const bytes = new Uint8Array(len);
  crypto.getRandomValues(bytes);
  for (let i = 0; i < len; i++) {
    result += chars[bytes[i] % chars.length];
  }
  return result;
}

/** 生成强密码（24 位，仅包含大小写字母与数字；更适合直接用于代理 URL userinfo） */
function generateStrongPassword(): string {
  const lower = "abcdefghijklmnopqrstuvwxyz";
  const upper = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
  const digits = "0123456789";
  const all = lower + upper + digits;
  const len = 24;

  const randIndex = (mod: number): number => {
    const b = new Uint32Array(1);
    crypto.getRandomValues(b);
    return b[0] % mod;
  };

  // 确保至少包含每种类型各一个
  const required = [
    lower[randIndex(lower.length)],
    upper[randIndex(upper.length)],
    digits[randIndex(digits.length)]
  ];

  const bytes = new Uint8Array(len - required.length);
  crypto.getRandomValues(bytes);
  const rest: string[] = [];
  for (let i = 0; i < bytes.length; i++) {
    rest.push(all[bytes[i] % all.length]);
  }

  // 合并并打乱顺序
  const combined = [...required, ...rest];
  for (let i = combined.length - 1; i > 0; i--) {
    const j = randIndex(i + 1);
    [combined[i], combined[j]] = [combined[j], combined[i]];
  }
  return combined.join("");
}

/** 生成新的代理认证凭据 */
export function generateProxyAuth(): ProxyAuth {
  return {
    enabled: false,
    username: generateRandomUsername(),
    password: generateStrongPassword()
  };
}

const DEFAULT_STATE: State = {
  version: 1,
  settings: {
    bindAddress: "127.0.0.1",
    allowLan: false,
    logLevel: "info",
    baseMixedPort: 30001,
    baseControllerPort: 40001,
    maxLogLines: 800,
    healthCheckIntervalSec: 60,
    healthCheckUrl: "http://www.gstatic.com/generate_204",
    exportHost: "",
    proxyAuth: generateProxyAuth()
  },
  subscriptions: [],
  instances: []
};

export class Storage {
  readonly dataDir: string;
  readonly dbPath: string;
  private db: Database | null = null;

  constructor(dataDir: string) {
    this.dataDir = dataDir;
    this.dbPath = join(dataDir, "state.sqlite");
  }

  private openDb(): Database {
    if (this.db) return this.db;
    const db = new Database(this.dbPath);
    db.exec("PRAGMA journal_mode = WAL;");
    db.exec("PRAGMA foreign_keys = ON;");
    db.exec("CREATE TABLE IF NOT EXISTS kv (key TEXT PRIMARY KEY, value TEXT NOT NULL);");
    this.db = db;
    return db;
  }

  async ensure(): Promise<void> {
    await mkdir(this.dataDir, { recursive: true });
    await mkdir(join(this.dataDir, "instances"), { recursive: true });
    this.openDb();
  }

  private normalizeState(maybe: unknown): State {
    const fallback = structuredClone(DEFAULT_STATE);
    if (!maybe || typeof maybe !== "object") return fallback;
    const parsed = maybe as any;
    if (parsed?.version !== 1) return fallback;

    const settings = parsed?.settings ?? {};

    // 处理 proxyAuth：如果旧数据没有或格式不对，生成新的
    let proxyAuth = fallback.settings.proxyAuth;
    if (settings.proxyAuth && typeof settings.proxyAuth === "object") {
      const pa = settings.proxyAuth;
      proxyAuth = {
        enabled: typeof pa.enabled === "boolean" ? pa.enabled : false,
        username: typeof pa.username === "string" && pa.username ? pa.username : generateProxyAuth().username,
        password: typeof pa.password === "string" && pa.password ? pa.password : generateProxyAuth().password
      };
    }

    const mergedSettings = {
      ...fallback.settings,
      bindAddress: typeof settings.bindAddress === "string" && settings.bindAddress ? settings.bindAddress : fallback.settings.bindAddress,
      allowLan: typeof settings.allowLan === "boolean" ? settings.allowLan : fallback.settings.allowLan,
      logLevel: ["silent", "error", "warning", "info", "debug"].includes(settings.logLevel) ? settings.logLevel : fallback.settings.logLevel,
      baseMixedPort: Number.isFinite(settings.baseMixedPort) ? Number(settings.baseMixedPort) : fallback.settings.baseMixedPort,
      baseControllerPort: Number.isFinite(settings.baseControllerPort) ? Number(settings.baseControllerPort) : fallback.settings.baseControllerPort,
      maxLogLines: Number.isFinite(settings.maxLogLines) ? Number(settings.maxLogLines) : fallback.settings.maxLogLines,
      healthCheckIntervalSec:
        Number.isFinite(settings.healthCheckIntervalSec) && Number(settings.healthCheckIntervalSec) >= 0
          ? Number(settings.healthCheckIntervalSec)
          : fallback.settings.healthCheckIntervalSec,
      healthCheckUrl:
        typeof settings.healthCheckUrl === "string" && settings.healthCheckUrl.trim()
          ? settings.healthCheckUrl.trim()
          : fallback.settings.healthCheckUrl,
      exportHost: typeof settings.exportHost === "string" ? settings.exportHost.trim() : fallback.settings.exportHost,
      proxyAuth
    };

    const subscriptions: Subscription[] = Array.isArray(parsed.subscriptions)
      ? parsed.subscriptions.filter((s: unknown) => !!s && typeof s === "object")
      : [];

    const instances: Instance[] = Array.isArray(parsed.instances)
      ? parsed.instances
          .filter((i: unknown) => !!i && typeof i === "object")
          .map((i: any) => ({
            ...i,
            autoSwitch: typeof i.autoSwitch === "boolean" ? i.autoSwitch : false
          }))
      : [];

    return {
      version: 1,
      settings: mergedSettings,
      subscriptions,
      instances
    };
  }

  private tryLoadLegacyStateJson(): Promise<State | null> {
    const legacyPath = join(this.dataDir, "state.json");
    if (!existsSync(legacyPath)) return Promise.resolve(null);
    return readFile(legacyPath, "utf8")
      .then((raw) => this.normalizeState(JSON.parse(raw)))
      .catch(() => null);
  }

  async loadState(): Promise<State> {
    await this.ensure();
    const db = this.openDb();

    const row = db.query("SELECT value FROM kv WHERE key = ?").get("state") as { value: string } | null;
    if (row?.value) {
      try {
        return this.normalizeState(JSON.parse(row.value));
      } catch {
        // fallthrough
      }
    }

    const legacy = await this.tryLoadLegacyStateJson();
    const initial = legacy ?? structuredClone(DEFAULT_STATE);
    db.query("INSERT OR REPLACE INTO kv (key, value) VALUES (?, ?)").run("state", JSON.stringify(initial));

    if (legacy) {
      try {
        await rename(join(this.dataDir, "state.json"), join(this.dataDir, "state.json.bak"));
      } catch {
        // ignore
      }
    }

    return initial;
  }

  async saveState(state: State): Promise<void> {
    await this.ensure();
    const db = this.openDb();
    db.query("INSERT OR REPLACE INTO kv (key, value) VALUES (?, ?)").run("state", JSON.stringify(state));
  }

  getKv(key: string): string | null {
    const db = this.openDb();
    const row = db.query("SELECT value FROM kv WHERE key = ?").get(key) as { value: string } | null;
    return row?.value ?? null;
  }

  setKv(key: string, value: string): void {
    const db = this.openDb();
    db.query("INSERT OR REPLACE INTO kv (key, value) VALUES (?, ?)").run(key, value);
  }

  deleteKv(key: string): void {
    const db = this.openDb();
    db.query("DELETE FROM kv WHERE key = ?").run(key);
  }

  getJson<T>(key: string): T | null {
    const raw = this.getKv(key);
    if (!raw) return null;
    try {
      return JSON.parse(raw) as T;
    } catch {
      return null;
    }
  }

  setJson(key: string, value: unknown): void {
    this.setKv(key, JSON.stringify(value));
  }

  deleteJson(key: string): void {
    this.deleteKv(key);
  }
}

File: api/src/server/storage.go
package server

import (
	"database/sql"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sync"

	_ "modernc.org/sqlite"
)

const stateVersion = 1

type Storage struct {
	dataDir string
	dbPath  string

	mu sync.Mutex
	db *sql.DB
}

func defaultState() State {
	return State{
		Version: stateVersion,
		Settings: Settings{
			BindAddress:            "127.0.0.1",
			AllowLan:               false,
			LogLevel:               "info",
			BaseMixedPort:          30001,
			BaseControllerPort:     40001,
			MaxLogLines:            800,
			HealthCheckIntervalSec: 60,
			SubscriptionRefreshMin: 0,
			HealthCheckURL:         "http://www.gstatic.com/generate_204",
			ExportHost:             "",
			ProxyAuth:              generateProxyAuth(),
		},
		Subscriptions: []Subscription{},
		Instances:     []Instance{},
	}
}

func NewStorage(dataDir string) *Storage {
	return &Storage{
		dataDir: dataDir,
		dbPath:  filepath.Join(dataDir, "state.sqlite"),
	}
}

func (s *Storage) ensureDirs() error {
	if err := os.MkdirAll(s.dataDir, 0o755); err != nil {
		return err
	}
	if err := os.MkdirAll(filepath.Join(s.dataDir, "instances"), 0o755); err != nil {
		return err
	}
	if err := os.MkdirAll(filepath.Join(s.dataDir, "subscriptions"), 0o755); err != nil {
		return err
	}
	return nil
}

func (s *Storage) openDBLocked() (*sql.DB, error) {
	if s.db != nil {
		return s.db, nil
	}
	if err := s.ensureDirs(); err != nil {
		return nil, err
	}
	db, err := sql.Open("sqlite", s.dbPath)
	if err != nil {
		return nil, err
	}
	stmts := []string{
		"PRAGMA journal_mode = WAL;",
		"PRAGMA foreign_keys = ON;",
		"PRAGMA busy_timeout = 5000;",
		"CREATE TABLE IF NOT EXISTS kv (key TEXT PRIMARY KEY, value TEXT NOT NULL);",
	}
	for _, stmt := range stmts {
		if _, err := db.Exec(stmt); err != nil {
			_ = db.Close()
			return nil, err
		}
	}
	s.db = db
	return db, nil
}

func (s *Storage) withDB(fn func(*sql.DB) error) error {
	s.mu.Lock()
	db, err := s.openDBLocked()
	s.mu.Unlock()
	if err != nil {
		return err
	}
	return fn(db)
}

func (s *Storage) GetKV(key string) (string, error) {
	var out string
	err := s.withDB(func(db *sql.DB) error {
		row := db.QueryRow("SELECT value FROM kv WHERE key = ?", key)
		if err := row.Scan(&out); err != nil {
			if errors.Is(err, sql.ErrNoRows) {
				out = ""
				return nil
			}
			return err
		}
		return nil
	})
	if err != nil {
		return "", err
	}
	return out, nil
}

func (s *Storage) SetKV(key, value string) error {
	return s.withDB(func(db *sql.DB) error {
		_, err := db.Exec("INSERT OR REPLACE INTO kv (key, value) VALUES (?, ?)", key, value)
		return err
	})
}

func (s *Storage) DeleteKV(key string) error {
	return s.withDB(func(db *sql.DB) error {
		_, err := db.Exec("DELETE FROM kv WHERE key = ?", key)
		return err
	})
}

func (s *Storage) GetJSON(key string, out any) error {
	raw, err := s.GetKV(key)
	if err != nil {
		return err
	}
	if raw == "" {
		return sql.ErrNoRows
	}
	if err := json.Unmarshal([]byte(raw), out); err != nil {
		return err
	}
	return nil
}

func (s *Storage) SetJSON(key string, in any) error {
	b, err := json.Marshal(in)
	if err != nil {
		return err
	}
	return s.SetKV(key, string(b))
}

func normalizeState(st State) State {
	fallback := defaultState()
	if st.Version != stateVersion {
		return fallback
	}
	out := st
	if out.Settings.BindAddress == "" {
		out.Settings.BindAddress = fallback.Settings.BindAddress
	}
	switch out.Settings.LogLevel {
	case "silent", "error", "warning", "info", "debug":
	default:
		out.Settings.LogLevel = fallback.Settings.LogLevel
	}
	if out.Settings.BaseMixedPort <= 0 {
		out.Settings.BaseMixedPort = fallback.Settings.BaseMixedPort
	}
	if out.Settings.BaseControllerPort <= 0 {
		out.Settings.BaseControllerPort = fallback.Settings.BaseControllerPort
	}
	if out.Settings.MaxLogLines <= 0 {
		out.Settings.MaxLogLines = fallback.Settings.MaxLogLines
	}
	if out.Settings.HealthCheckIntervalSec < 0 {
		out.Settings.HealthCheckIntervalSec = fallback.Settings.HealthCheckIntervalSec
	}
	if out.Settings.SubscriptionRefreshMin < 0 {
		out.Settings.SubscriptionRefreshMin = fallback.Settings.SubscriptionRefreshMin
	}
	if out.Settings.HealthCheckURL == "" {
		out.Settings.HealthCheckURL = fallback.Settings.HealthCheckURL
	}
	if out.Settings.ProxyAuth.Username == "" || out.Settings.ProxyAuth.Password == "" {
		next := generateProxyAuth()
		next.Enabled = out.Settings.ProxyAuth.Enabled
		out.Settings.ProxyAuth = next
	}
	if out.Subscriptions == nil {
		out.Subscriptions = []Subscription{}
	}
	if out.Instances == nil {
		out.Instances = []Instance{}
	}
	for i := range out.Instances {
		_ = out.Instances[i].AutoSwitch
	}
	return out
}

func (s *Storage) LoadState() (State, error) {
	if err := s.ensureDirs(); err != nil {
		return State{}, err
	}
	var st State
	if err := s.GetJSON("state", &st); err == nil {
		return normalizeState(st), nil
	}

	legacyPath := filepath.Join(s.dataDir, "state.json")
	if b, err := os.ReadFile(legacyPath); err == nil {
		if err := json.Unmarshal(b, &st); err == nil {
			norm := normalizeState(st)
			if err := s.SetJSON("state", norm); err != nil {
				return State{}, err
			}
			_ = os.Rename(legacyPath, legacyPath+".bak")
			return norm, nil
		}
	}

	initial := defaultState()
	if err := s.SetJSON("state", initial); err != nil {
		return State{}, err
	}
	return initial, nil
}

func (s *Storage) SaveState(st State) error {
	return s.SetJSON("state", st)
}

func (s *Storage) Close() error {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.db == nil {
		return nil
	}
	err := s.db.Close()
	s.db = nil
	return err
}

func (s *Storage) String() string {
	return fmt.Sprintf("Storage{%s}", s.dbPath)
}
